Generating physically-consistent high-resolution climate data with hard-constrained neural networks

aniket2405 opened the pull request:
# Pull Request for #52 

## Description

This PR introduces functionality for generating physically-consistent, high-resolution climate data using hard-constrained neural networks. The primary addition is the implementation of hard constraints, ensuring that generated climate data adheres to fundamental physical laws. This constraint mechanism is modular and can be applied to any model within this repository or following the same format.

## Motivation and Context
Accurate climate modeling demands that generated data respect physical laws (e.g., conservation of mass, energy). By adding hard constraints, models in this repository will produce outputs that are both high-resolution and physically plausible, improving their utility for climate science applications.

This implementation follows the methodology outlined in the paper:
[Generating Physically-Consistent High-Resolution Climate Data with Hard-Constrained Neural Networks](https://arxiv.org/pdf/2208.05424).
The approach is inspired by and adapted from the [RolnickLab's constrained-downscaling source code](https://github.com/RolnickLab/constrained-downscaling).

Fixes #52 

## How Has This Been Tested?

I have written unit tests to verify the correct implementation of the hard-constrained neural network logic. These tests focus on:

Constraint Enforcement
Model Compatibility

- [x] Yes

_If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?_

- [ ] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

To be reviewed: @jacobbieker

jacobbieker commented:
Quick scan this is looking good! Thanks for this! Tag me as a reviewer when the tests are not as flaky, and you feel it is good to go. Thanks for working on this!

aniket2405 commented:
Hi @jacobbieker 
This took a long time, but learnt a whole lot!
The tests are no more flaky and seem to be working fine - updated the physical constraints.
Please review the implementation, and I'd be happy to discuss how this can be improved. Thanks!

jacobbieker commented on graph_weather/models/forecast.py:
```diff
@@ -29,6 +30,9 @@ def __init__(
         norm_type: str = "LayerNorm",
         use_checkpointing: bool = False,
+        # New constraint parameters
+        constraint_type: str = "additive",  # "additive", "multiplicative", or "softmax"
+        apply_constraints: bool = True,
```
Maybe combining these two parameters make more sense, and default to "none"/`None`
```suggestion
        constraint_type: str = "additive",  # "additive", "multiplicative", "softmax", or "none"
```

aniket2405 replied:
Ack, made the change

jacobbieker commented on graph_weather/models/forecast.py:
```diff
@@ -111,4 +172,22 @@ def forward(self, features: torch.Tensor) -> torch.Tensor:
+            # (Original features has shape [B, num_nodes, feature_dim])
+            lr = features[..., : self.feature_dim]  # shape: [B, num_nodes, feature_dim]
+            # Convert from node format to grid format using the grid_shape computed in __init__
+            lr = lr.permute(0, 2, 1)  # becomes [B, feature_dim, num_nodes]
```
I have a small preference for using `einops` for these kinds of changes, just to make it more readable.

jacobbieker commented on graph_weather/models/forecast.py:
```diff
@@ -111,4 +172,22 @@ def forward(self, features: torch.Tensor) -> torch.Tensor:
+            # Convert from node format to grid format using the grid_shape computed in __init__
+            lr = lr.permute(0, 2, 1)  # becomes [B, feature_dim, num_nodes]
+
+            lr = lr.view(batch_size, self.feature_dim, self.grid_shape[0], self.grid_shape[1])
```
Same here, I don't think using einops should be any different than `view` in terms of memory usage, and could be a bit more readable.

aniket2405 replied:
Done

jacobbieker commented on tests/test_model.py:
```diff
@@ -124,13 +124,12 @@ def test_forecaster():
-    # Add in auxiliary features
-    features = torch.randn((1, len(lat_lons), 78 + 24))
 
+    model = GraphWeatherForecaster(lat_lons, feature_dim=78, output_dim=78, apply_constraints=False)
```
Ideally, the changes for this PR shouldn't require changing this test, so having the constraints default to False, or the `constraint_type="none"` as mentioned above might be a better way of doing it.

aniket2405 replied:
Yes, using `constraint_type="none"` prevent any need for change in the existing unit tests. Good find!

jacobbieker commented on tests/test_model.py:
```diff
@@ -332,3 +348,37 @@ def test_wrapper_meta_model():
     assert out.size() == big_features.size()
+
+
+def test_constrained_forecast():
```
Nice test! I would have it do each type of constraint, as each of its own unit test.

jacobbieker reviewed (changes requested):
This is looking really good! A few small nits on trying to use `einops` to be more readable, and a small change to how the configuration is set.

aniket2405 commented:
Thanks @jacobbieker for the review and comments. I've gone through them and made the changes, please have a review now and let me know how this looks!

jacobbieker commented on graph_weather/models/forecast.py:
```diff
@@ -29,6 +31,7 @@ def __init__(
         hidden_layers_decoder: int = 2,
         norm_type: str = "LayerNorm",
         use_checkpointing: bool = False,
+        constraint_type: str = "additive",  # "additive", "multiplicative", "softmax" or "none"
```
I would default to "none"
```suggestion
        constraint_type: str = "none",  # "additive", "multiplicative", "softmax" or "none"
```

aniket2405 replied:
Ack, forgot to update this.

jacobbieker commented on graph_weather/models/forecast.py:
```diff
@@ -98,6 +114,49 @@ def __init__(
         )
 
+        # Add physical constraint layer
+        self.constraint = PhysicalConstraintLayer(
```
This should only be made if the constraint is not "none"

jacobbieker reviewed (changes requested):
This looks good! Two minor changes, then I think its ready to merge!

aniket2405 commented:
hi @jacobbieker 
Updated the code with your suggestions.

There's something else - 

When there is no physical constraint by default, the `NormalizedMSELoss` throws an error.

When we use a constraint type - the constraint layer converts the decoder’s output from node format into a grid format whose spatial dimensions are determined by the unique latitudes and longitudes. In practice, the transformation done in the constraint layer (and the subsequent operations) leads to a tensor with one value per unique latitude when we eventually reduce along one axis.

That means the expected weight tensor should have one weight per unique latitude rather than one per (lat, lon) pair.

When we set the constraint type to "none", the physical constraint isn’t applied, so the output is not further transformed by that layer. However, our loss function still expects the weight tensor to broadcast correctly with an error tensor whose spatial dimension now corresponds to the unique latitudes from the grid conversion rather than the full lat–lon grid. 

Since we originally built the weight tensor using every (lat, lon) pair, we end up with a size mismatch, which causes a runtime error when trying to expand the weights.

To resolve this, we must change the loss code to compute the weights from unique latitudes instead of every lat–lon pair.

I have updated these changes and the unit tests are passing now.

Let me know if there is a better way to resolve this which you'd like me to try.

jacobbieker commented:
Yeah, I think that is reasonable. Could you add a test for an irregular grid? Maybe just something with `pytest.parameterize` or something, but most of the tests are with regulare lat/lon grids, and we want to make sure it also works with irregular ones. So some type of test where every input lat and input lon is unique? If it passes that too, then I think its good to go then!

jacobbieker commented:
Yeah, fair enough, making the physical constraint layer work with irregular grids can be a later issue. The current implementation in the repo does work with irregular grids though, so would prefer one check to then ensure the current irregular grid support doesn't disappear with this. Here is a test that the `main` branch passes, and should still pass with this PR: 

```
def test_forecaster_and_loss_irregular():
    lat_lons = []
    for lat in range(-90, 90, 5):
        for lon in range(0, 360, 5):
            lat_lons.append((lat, lon))
    # Jitter the lat and lons
    lat_lons = [(lat + np.random.random(), lon + np.random.random()) for lat, lon in lat_lons]
    criterion = NormalizedMSELoss(lat_lons=lat_lons, feature_variance=torch.randn((78,)))
    model = GraphWeatherForecaster(lat_lons)
    # Add in auxiliary features
    features = torch.randn((2, len(lat_lons), 78 + 24))

    out = model(features)
    loss = criterion(out, torch.rand(out.shape))
    assert not torch.isnan(loss)
    assert not torch.isnan(out).any()
    assert not torch.isnan(out).any()
    loss.backward()
```

If this passes with the current PR, then I'm fine merging it.

aniket2405 commented:
hi @jacobbieker 
This test passes after a slight modification, rearranging the input from graph to grid format only when physical constraints are applied. 

This ensures irregular grids work just like before when physical_constrains are "none".

Ensured that all unit tests pass now. We can go forward and merge this, if everything looks good to you too.

Q. Why is this unit test -  `test_forecaster_and_loss_irregular` - not part of the test suite in the code?

jacobbieker commented:
> Q. Why is this unit test - `test_forecaster_and_loss_irregular` - not part of the test suite in the code?

It is now, I thought the test suite already had this kind of test, I just added it as part of #138

jacobbieker reviewed (approved):
LGTM! I might have accidentally caused the merge conflict, but if you can fix that, let's merge it! Thanks for all this work!

aniket2405 commented:
Oh, right.
Looks like this was merged today, didn't update the codebase. Thanks!

jacobbieker commented:
Thanks for fixing that! Happy to merge it now if you are?

aniket2405 commented:
Yeah, I'm happy to merge.
There are some pending ruff pre-commit errors/warnings, though.
Don't we need to fix them before merging?

jacobbieker commented:
The pre-commit checks are a bit broken right now, so I wouldn't worry too much about that. I'll merge it now then.

aniket2405 commented:
Oh alright, thanks for merging this!
